galaxy_ng/app/migrations/0026_move_perms_to_roles.py

```diff
@@ -0,0 +1,136 @@
+    ), "galaxy.synclist_owner"),
+]
+
+
+def locked_perm_in_permissions(locked_perm, permissions):
+    # locked_perm: Permission object
+    # permissions: queryset of Permission objects
+    for perm in permissions:
+        if perm.id == locked_perm.id:
+            return True
+    return False
+
+
+def remove_locked_perms_from_permissions(perms_to_remove, permissions):
+    # permissions_to_remove: list of Permission objects to remove
+    # permissions: queryset of Permission objects
+    for p in perms_to_remove:
+        permissions = permissions.exclude(id=p.id)
+    return permissions
+
+
+def move_permissions_to_roles(apps, schema_editor):
+    Group = apps.get_model("galaxy", "Group")
+    GroupRole = apps.get_model("core", "GroupRole")
+    Permission = apps.get_model("auth", "Permission")
+    Role = apps.get_model("core", "Role")
+
+    group_roles = []
+
+    for i, group in enumerate(Group.objects.all()):
+        if group.name == "system:partner-engineers":
+            continue
+        permissions = group.permissions.all()
+        perms_to_remove = []
+
+        # Use Galaxy locked Roles where possible
+        for locked_perm_names, locked_rolename in GALAXY_LOCKED_ROLE_TRANSLATOR:
+            locked_perms = [Permission.objects.filter(
+                content_type__app_label=app_label, codename=codename
+            ).first() for app_label, codename in locked_perm_names]
+
+            if all(locked_perms):
+                # compare locked role perms to perms
+                for k, locked_perm in enumerate(locked_perms):
+                    if locked_perm_in_permissions(locked_perm, permissions):
+                        # Is that all of the permissions in this locked_role?
+                        if (k + 1) == len(locked_perms):
+                            # add current locked_role to list to add to current group
+                            locked_role, _ = Role.objects.get_or_create(
+                                name=locked_rolename,
+                                defaults={"locked": True}
+                            )
+                            group_roles.append(GroupRole(group=group, role=locked_role))
+
+                            # add permissions from locked_perms to list to be removed
+                            for p in locked_perms:
+                                perms_to_remove.append(p)
+                        continue
+                    else:
+                        break
+                # remove permissions covered by locked_roles from the groups permission list
+                permissions = remove_locked_perms_from_permissions(perms_to_remove, permissions)
+
+        # Create custom role for current Group with permissions not in Galaxy locked roles
+        if len(permissions) > 0:
+            role, _ = Role.objects.get_or_create(name=f"{group.name}_role")
+            for permission in permissions:
+                role.permissions.add(permission)
```

Comment:
Maybe this can work:
```suggestion
            role.permissions.set(permissions)
```